src/foam/core/Action.js

```diff
@@ -134,59 +134,58 @@ foam.CLASS({
   ],
 
   methods: [
-    function andSlots(a, b) {
-      return this.ExpressionSlot.create({
-        args: [ a, b ],
-        code: function(a, b) {
-          return a && b;
-        }
-      });
-    },
 
-    function andSlotAndPromise(slot, promise) {
-      return this.andSlots(slot, this.PromiseSlot.create({
-        promise: promise
-      }));
+    function hasPermissions_(x, property) {
+      if ( ! this.permissionRequired || ! x.auth ) return Promise.resolve(true);
+      return Promise.all(this.permissionConfig[property].map(p => {
+        return x.auth.check(null, p);
+      })).then(arr => arr.every(b => b));
     },
 
-    function checkPermission(x, permission) {
-      if ( ! this.permissionConfig || ! x.auth ) return Promise.resolve(true);
-      return x.auth.check(null, permission);
-    },
-
-    function isEnabledFor(data) {
-      return this.isEnabled
-        ? foam.Function.withArgs(this.isEnabled, data)
-        : true;
-    },
-
-    function createIsEnabled$(data$) {
-      return this.createSlot_(this.isEnabled, 'enabled', data$);
+    function createPermissionlessCheckFor_(data$, property) {
+      if ( ! this[property] ) return foam.core.ConstantSlot.create({value: true});
+      return foam.core.ExpressionSlot.create({
+        obj$: data$,
+        code: this[property]
+      });
     },
 
-    function isAvailableFor(data) {
-      return this.isAvailable
-        ? foam.Function.withArgs(this.isAvailable, data)
-        : true;
+    function createIsEnabled$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
+            promise: this.hasPermissions_(x, 'enabled')
+          }),
+          this.createPermissionlessCheckFor_(data$, 'isEnabled')
+        ]
+      }).map(arr => arr.every(b => b));
     },
 
-    function createIsAvailable$(data$) {
-      return this.createSlot_(this.isAvailable, 'available', data$);
+    function createIsAvailable$(x, data$) {
+      return foam.core.ArraySlot.create({
+        slots: [
+          foam.core.PromiseSlot.create({
+            promise: this.hasPermissions_(x, 'available')
+          }),
+          this.createPermissionlessCheckFor_(data$, 'isAvailable')
+        ]
+      }).map(arr => arr.every(b => b));
     },
 
-    async function maybeCall(ctx, data) {
+    function maybeCall(ctx, data) {
       var data$ = this.ConstantSlot.create({ value: data });
-      if (
-        this.isEnabledFor(data) &&
-        this.isAvailableFor(data) &&
-        await this.hasPermissions_('enabled', data$) &&
-        await this.hasPermissions_('available', data$)
-      ) {
+      Promise.all([
```

Comment:
Let's add a comment here noting that because the action is called asynchronously, if it attempts to open any windows with window.open then it will get rejected by the browser's popup blocker.

As you can only do window.open synchronously from an onclick event handler.

It would be extra nice but not as elegant if we avoided the async path in the case where we don't have permissions or we don't have an auth service to check again.

Reply:
I agree.